### ISSUE TITLE
windows: add MSI installer

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#595]
A number of people have requested MSI installers.

This is a tracking bug for that.

<img src="https://frontapp.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_2a89d)

**Comment by bradfitz** [comment]
@alexbrainman, do you know anything about building MSIs?

**Comment by alexbrainman** [comment]
> @alexbrainman, do you know anything about building MSIs?

I know as much as you do. Whatever I contributed to Go installer over the years. So not much.

From what I remember, MSI build tools are running on Windows. Does that concern you?

I also recently saw this

https://github.com/WireGuard/wireguard-windows/tree/master/installer

might be useful.

Why do we need MSI installer? What is wrong with current installer?

Alex

**Comment by bradfitz** [comment]
> Why do we need MSI installer? What is wrong with current installer?

Primarily because many organizations regularly request MSI installers. For some it's even a requirement and they won't use an EXE.

MSI is a common interface that hooks into many fleet management tools with standard ways to install/upgrade/uninstall/query. See https://superuser.com/questions/38221/install-software-choose-msi-or-exe or https://stackoverflow.com/questions/1284825/windows-application-installer-frameworks or https://stackoverflow.com/questions/49624070/how-to-create-windows-installer/49632260#49632260 for some background on why MSIs are preferred over arbitrary side-effect-producing EXEs.

https://wiki.gnome.org/msitools suggests you don't need Windows to build MSI files.

**Comment by alexbrainman** [comment]
> Primarily because many organizations regularly request MSI installers. For some it's even a requirement and they won't use an EXE.

Fair enough. And thanks for the links.

I don't think it will be particularly difficult to build MSI installer for Tailscale. Someone just need to do it. And we need to decide what to do about wintun driver before we start changing the installer. Maybe we could just use original wintun driver as is. It comes with its own MSI installer

https://git.zx2c4.com/wintun/about/#installation

Alex

**Comment by AdrianFletcher** [comment]
Would also love an MSI installer. I can specially bundle the EXE installer using a custom InTune package for deployment for example - but an MSI is just click and forget.

I believe the official Wireguard for Windows installer is MSI - there may be some good information in the repo?

https://github.com/WireGuard/wireguard-windows

**Comment by DentonGentry** [comment]
Noting for the future, though I've not looked at these much:
- https://github.com/mh-cbon/go-msi
- https://www.npmjs.com/package/msi-packager

**Comment by DentonGentry** [comment]
We're looking seriously at https://wixtoolset.org/ to create an MSI installer. It nominally only runs on Windows, but people have been successful in getting it to run in a Linux system: https://github.com/suchja/wix-toolset

**Comment by zx2c4** [comment]
> I believe the official Wireguard for Windows installer is MSI - there may be some good information in the repo?

The WiX XML itself: https://git.zx2c4.com/wireguard-windows/tree/installer/wireguard.wxs
Some custom actions we add to it: https://git.zx2c4.com/wireguard-windows/tree/installer/customactions.c
The way in which it's built: https://git.zx2c4.com/wireguard-windows/tree/installer/build.bat

**Comment by ndom91** [comment]
What is the current suggested way to roll out tailscale in an enterprise windows environment? 

I really want to push my managers on tailscale, but without an easy way to roll this out to client machines it's going to be a hard sell. Unfortunately we don't have any third party software management service, we've been successfully using MSI's in GPO's for everything so far.

**Comment by DentonGentry** [comment]
Most larger deployments so far have used Microsoft SCCM or similar infrastructure.

I expect that we will repackage the Windows installer as MSI, it is just a question of engineering scheduling.

**Comment by ndom91** [comment]
Okay makes sense. Yeah an MSI installer would make tailscale deployment soo much easier for all the admins out there :)

**Comment by DentonGentry** [comment]
We decided not to try to yeet MSI out the door with the 1.18 release. Instead we'll phase it in on the unstable track first https://pkgs.tailscale.com/unstable/, to get some exposure with people willing to try it early.

**Comment by DentonGentry** [comment]
MSI installers are available on https://pkgs.tailscale.com/unstable/#windows for the 1.21.x unstable release.

**Comment by AdrianFletcher** [comment]
Noticed that the changelog says 1.22 has a supported MSI installer - is there a link to the download somewhere? Struggling to find it.

**Comment by bradfitz** [comment]
https://pkgs.tailscale.com/stable/ has it.

**Comment by AdrianFletcher** [comment]
Brilliant, thanks.

**Comment by DentonGentry** [comment]
As an MSI installer was delivered in the 1.22 release, I'm closing this as done.

If you notice issues with the MSI installer please do not comment on this issue, as we don't monitor closed bugs. Please file a new issue, or send details to support@tailscale.com if something is not suitable to post publicly.
